zaza/openstack/charm_tests/nova/tests.py

```diff
@@ -141,6 +142,107 @@ def test_930_check_virsh_default_network(self):
                 model_name=self.model_name)
             self.assertFalse(int(run['Code']) == 0)
 
+    def test_1000_ceph_broker_message_marked_done(self):
+        """Test if ceph broker messages were exchanged correctly."""
+        backend = zaza.model.get_application_config(
+            self.application_name)['libvirt-image-backend']
+        rbd_enabled = backend['value'].lower() == "rbd"
+        if not rbd_enabled:
+            raise unittest.SkipTest(
+                'RBD backend is not enabled, skipping test')
+
+        ceph_sentries = zaza.model.get_units(
+            'ceph-mon', model_name=self.model_name)
+        nova_sentries = zaza.model.get_units(
+            'nova-compute', model_name=self.model_name)
+        for nova_sentry in nova_sentries:
+            nova_unit_name = nova_sentry.entity_id
+
+            marked_done = False
+            for ceph_sentry in ceph_sentries:
+                ceph_unit_name = ceph_sentry.entity_id
+                logging.info('Checking ceph broker messaging between '
+                             'units {} and {}'.format(nova_unit_name,
+                                                      ceph_unit_name))
+                output = zaza_juju.get_relation_from_unit(
+                    nova_unit_name, ceph_unit_name, 'client')
+                logging.info('Output of relation is: {}'.format(output))
+                broker_key = 'broker-rsp-{}'.format(
+                    nova_unit_name.replace('/', '-'))
+                if broker_key in output.keys():
+                    broker_rsp_string = output[broker_key]
+                    broker_rsp_dict = json.loads(broker_rsp_string)
+                    unit_number = nova_unit_name.partition('/')[2]
+                    key = "unit_{}_ceph_broker_action." \
+                          "nova_compute_restart".format(unit_number)
+                    command = 'cd hooks;python -c "' \
+                              'from charmhelpers.core.unitdata import kv;' \
+                              'kvstore = kv();' \
+                              'val = kvstore.get(key=\\"{}\\");' \
+                              'print(val)"'.format(key)
+                    output = zaza.model.run_on_unit(
+                        nova_unit_name, command).get('Stdout').strip()
+                    if broker_rsp_dict['request-id'] in output:
+                        logging.info(
+                            'Found request-id in output: '
+                            '{}'.format(broker_rsp_dict['request-id']))
+                        marked_done = True
+
+            if not marked_done:
+                msg = "Ceph broker message was not processed " \
+                      "successfully for {}".format(nova_unit_name)
+                raise Exception("Test failed: {}".format(msg))
```

Comment:
This looks a bit like the test is peeking a bit too much into the implementation of the effect being tested rather than block box testing the outcome (behaviours) of the the system-under-test.  i.e. the test probably ought to test "what it should do" with a particular setup, rather than "how is is implemented".  This would allow the implementation to change, but the behaviour to remain the same, which is what is important.  The issue with testing implementation at this testing level, is it 'locks in' the implementation.

I think the purpose is to check that the "nova" pool gets created, which is the purpose of the next test, but there may be something that I don't understand or I'm missing.

Reply:
removed that and now only checks if the pool has been created